python/instrumentation/openinference-instrumentation-boto/src/openinference/instrumentation/boto/__init__.py

```diff
@@ -0,0 +1,169 @@
+
+_MODULE = "botocore.client"
+__version__ = "0.1.0"
+_instruments = ("boto3 >= 1.28.57",)
+
+
+def _set_span_attribute(span: trace_api.Span, name: str, value: AttributeValue) -> None:
+    if value is not None:
+        if value != "":
+            span.set_attribute(name, value)
+    return
+
+
+class InstrumentedClient(BaseClient):  # type: ignore
+    """
+    Proxy class representing an instrumented boto client.
+    """
+
+    invoke_model: Callable[..., Any]
+    _unwrapped_invoke_model: Callable[..., Any]
+
+
+class BufferedStreamingBody(StreamingBody):  # type: ignore
+    _raw_stream: IO[bytes]
+
+    def __init__(self, raw_stream: IO[bytes], content_length: int) -> None:
+        super().__init__(raw_stream, content_length)
+        self._buffer: Optional[io.IOBase] = None
+
+    def read(self, amt: Optional[int] = None) -> bytes:
+        if self._buffer is None:
+            self._buffer = io.BytesIO(self._raw_stream.read())
+
+        output: bytes = self._buffer.read(amt)
+        return output
+
+    def reset(self) -> None:
+        # Reset the buffer to enable reading the stream again
+        if self._buffer is not None:
+            self._buffer.seek(0)
+
+
+def _client_creation_wrapper(tracer: Tracer) -> Callable[[CallableType], CallableType]:
+    def _client_wrapper(
+        wrapped: CallableType,
+        instance: Optional[Any],
+        args: Tuple[Any, ...],
+        kwargs: Dict[str, Any],
+    ) -> BaseClient:
+        @wraps(wrapped)
+        def create_instrumented_client(*args: Any, **kwargs: Any) -> BaseClient:
+            """Instruments and calls every function defined in TO_WRAP."""
+            client: BaseClient = wrapped(*args, **kwargs)
+            if context_api.get_value(_SUPPRESS_INSTRUMENTATION_KEY):
+                return client
+
+            call_signature = signature(wrapped)
+            bound_arguments = call_signature.bind(*args, **kwargs)
+            bound_arguments.apply_defaults()
+
+            if bound_arguments.arguments.get("service_name") == "bedrock-runtime":
+                client = cast(InstrumentedClient, client)
+                client._unwrapped_invoke_model = client.invoke_model
+                client.invoke_model = _model_invocation_wrapper(tracer)(client)
+                return client
+            return client
+
+        return create_instrumented_client(*args, **kwargs)
+
+    return _client_wrapper  # type: ignore
+
+
+def _model_invocation_wrapper(tracer: Tracer) -> Callable[[InstrumentedClient], Callable[..., Any]]:
+    def _invocation_wrapper(wrapped_client: InstrumentedClient) -> Callable[..., Any]:
+        """Instruments a bedrock client's `invoke_model` method."""
+
+        @wraps(wrapped_client.invoke_model)
+        def instrumented_response(*args: Any, **kwargs: Any) -> Dict[str, Any]:
+            with tracer.start_as_current_span("bedrock.invoke_model") as span:
+                response = wrapped_client._unwrapped_invoke_model(*args, **kwargs)
```

Comment:
suggestion: instead of currying, we can also wrap the bound method with `wrapt.ObjectProxy`.
```python
client.invoke_model = _InvokeModel(client.invoke_model, tracer)
```

```suggestion
class _InvokeModel(ObjectProxy):
    def __init__(self, wrapped, tracer) -> None:
        super().__init__(wrapped)
        self._self_tracer = tracer

    def __call__(self, *args: Any, **kwargs: Any) -> Dict[str, Any]:
        with self._self_tracer.start_as_current_span("bedrock.invoke_model") as span:
            response = self.__wrapped__(*args, **kwargs)
```

Reply:
I'm using currying here to hold a reference to the original client in the closure. This makes it easier to test and patch the original `invoke_model` method after instrumentation. Using an ObjectProxy as implemented here would make this harder to test